x-pack/legacy/plugins/triggers_actions_ui/np_ready/public/application/components/builtin_action_types/webhook.tsx

```diff
@@ -0,0 +1,471 @@
+            {
+              defaultMessage: 'URL is required.',
+            }
+          )
+        );
+      }
+      if (!action.config.method) {
+        errors.method.push(
+          i18n.translate(
+            'xpack.triggersActionsUI.sections.addAction.webhookAction.error.requiredMethodText',
+            {
+              defaultMessage: 'Method is required.',
+            }
+          )
+        );
+      }
+      if (!action.secrets.user) {
+        errors.user.push(
+          i18n.translate(
+            'xpack.triggersActionsUI.sections.addAction.webhookAction.error.requiredHostText',
+            {
+              defaultMessage: 'Username is required.',
+            }
+          )
+        );
+      }
+      if (!action.secrets.password) {
+        errors.password.push(
+          i18n.translate(
+            'xpack.triggersActionsUI.sections.addAction.webhookAction.error.requiredPasswordText',
+            {
+              defaultMessage: 'Password is required.',
+            }
+          )
+        );
+      }
+      return validationResult;
+    },
+    validateParams: (actionParams: any): ValidationResult => {
+      const validationResult = { errors: {} };
+      return validationResult;
+    },
+    actionConnectorFields: WebhookActionConnectorFields,
+    actionParamsFields: WebhookParamsFields,
+  };
+}
+
+const WebhookActionConnectorFields: React.FunctionComponent<ActionConnectorFieldsProps> = ({
+  action,
+  editActionConfig,
+  editActionSecrets,
+  errors,
+}) => {
+  const [headerKey, setHeaderKey] = useState<string>('');
+  const [headerValue, setHeaderValue] = useState<string>('');
+  const [hasHeaders, setHasHeaders] = useState<boolean>(false);
+
+  const { user, password } = action.secrets;
+  const { method, url, headers } = action.config;
+
+  editActionConfig('method', 'post'); // set method to POST by default
+
+  const headerErrors = {
+    keyHeader: new Array<string>(),
+    valueHeader: new Array<string>(),
+  };
+  if (!headerKey && headerValue) {
+    headerErrors.keyHeader.push(
+      i18n.translate(
+        'xpack.triggersActionsUI.sections.addAction.webhookAction.error.requiredHeaderKeyText',
+        {
+          defaultMessage: 'Header key is required.',
+        }
+      )
+    );
+  }
+  if (headerKey && !headerValue) {
+    headerErrors.valueHeader.push(
+      i18n.translate(
+        'xpack.triggersActionsUI.sections.addAction.webhookAction.error.requiredHeaderValueText',
+        {
+          defaultMessage: 'Header value is required.',
+        }
+      )
+    );
+  }
+  const hasHeaderErrors = headerErrors.keyHeader.length > 0 || headerErrors.valueHeader.length > 0;
+
+  function addHeader() {
+    if (headers && !!Object.keys(headers).find(key => key === headerKey)) {
+      return;
+    }
+    const updatedHeaders = headers
+      ? { ...headers, [headerKey]: headerValue }
+      : { [headerKey]: headerValue };
```

Comment:
I wonder if these property settings are vulnerable to prototype pollution - eg, what if `headerKey` is `__proto__`.  Is there some handy link to JS prototype pollution?  I have an email with a link to a Google presentation - but that's all I know of. https://docs.google.com/presentation/d/1OCrWbXxiH-WgoJf85h3puFJGylLIF-OV2R5uofsk4To/edit#slide=id.p

Reply:
renamed to **httpHeaderKey** and **httpHeaderValue**